### DISCUSSION TITLE
IServiceCollectionQuartzConfigurator.ScheduleJob does not trigger job

### DISCUSSION THREAD

**Discussion by ruwentos** [quartznet/quartznet#1135] [Q&A] [answered]
According to the [documentation ](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/microsoft-di-integration.html#di-aware-job-factories) I'm using **IServiceCollectionQuartzConfigurator.ScheduleJob** like this.

`services.AddQuartz(q =>
{
    q.SchedulerId = "CutOff-time-scheduler";

    q.UseMicrosoftDependencyInjectionScopedJobFactory(options =>
    {
        options.CreateScope = true;
        options.AllowDefaultConstructor = false;
    });
    
    q.ScheduleJob<SendEventIfMarketECutOffTimePassedJob>(trigger => trigger
        .WithIdentity("daily-trigger")
        .WithCronSchedule("0/15 * * * * ?")
        .WithDescription("triggers the 'SendEventIfMarketECutOffTimePassed' job every day at ")
        .StartNow()
    );
});`

My job code looks like this:

`public class SendEventIfMarketECutOffTimePassedJob : IJob
    {
        private readonly ITopicPublisher<GenericEvent> _topicPublisher;
        private IAssetRepository _assetRepository;
        private IMarketRepository _marketRepository;
        private ICalendarRepository _calendarRepository;

        public SendEventIfMarketECutOffTimePassedJob(ITopicPublisher<GenericEvent> topicPublisher, IAssetRepository assetRepository, IMarketRepository marketRepository,
            ICalendarRepository calendarRepository)
        {
            _topicPublisher = topicPublisher;
            _assetRepository = assetRepository;
            _marketRepository = marketRepository;
            _calendarRepository = calendarRepository;
        }

        public Task Execute(IJobExecutionContext context)
        {
            var cutOffTimeEvent = new GenericEvent
            {
                Id = Guid.NewGuid(),
                SendFromObject = nameof(SendEventIfMarketECutOffTimePassedJob),
                Payload = "no payload"
            };

            if (CutOffTimePassed())
            {
                var _ = _topicPublisher
                    .SendEventAsync(cutOffTimeEvent, EventTopicNamesEnum.Market_e_cut_off_time_passed.ToString())
                    .Result;
            }
            return Task.CompletedTask;
        }

        private bool CutOffTimePassed()
        {
            return true;
        }
    }`

The job never gets triggered but should every 15 seconds.

I'm using fresh installed version:
`<PackageReference Include="Quartz" Version="3.2.4" />
    <PackageReference Include="Quartz.Extensions.DependencyInjection" Version="3.2.4" />`

**Comment by lahma** [comment, marked as answer]
You need to either start the scheduler manually (by resolving it from services/DI in your hosted service or similar) or use the [hosting package](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/hosted-services-integration.html) to start it automatically after configured as so:

https://github.com/quartznet/quartznet/blob/73f2e3132fdec18140986141a99d3248f2e5f972/src/Quartz.Examples.AspNetCore/Startup.cs#L224-L229

**Comment by ruza-kavanka** [comment]
I face the same issue. I'm trying to get sheduler instance by resolving DI from services, but there is no service registered with IScheduler. AddQuartzServer method is present in my code:

`services.Configure<QuartzOptions>(configuration.GetSection("Quartz"));
			services.AddQuartz(q=>{
				q.SchedulerId = "MDI scheduler";
				q.UseMicrosoftDependencyInjectionScopedJobFactory(jfo=>{
					jfo.AllowDefaultConstructor = true;
				});
			});
			services.AddQuartzServer(op=>{
				op.WaitForJobsToComplete = true;
			});
			services.AddSingleton<IQuartzJobFactory, QuartzJobFactory>();`

When I try to create scheduler instance by StdSchedulerFactory, the trigger doesn't fire.

**Reply by lahma** [reply]
There should be a `ISchedulerFactory` registered, you can call `IScheduler scheduler = await schedulerFactory.GetScheduler();` after you have resolved the scheduler factory. Enabling logging usually tells better what's happening.

**Reply by ruza-kavanka** [reply]
Thank you very much for reply, now I'm one step further - ISchedulerFactory is registered in services and I can get it with DI, perfect! 
But :-) my IJob class contains constructor with 2 parameters and they are supposed to be filled by DI. I have installed Quartz.Extensions.DependencyInjection and Microsoft.Extensions.DependencyInjection packages, and I have registered both classes for parameters to be filled by DI correctly. However - Quartz cannot create instance of my IJob, output error is: "Problem instantiating class 'Linet.MDI.Common.QuartzSalesRepsSynchroJob: Cannot instantiate type which has no empty constructor"
I tried both:
`services.AddQuartz(q=>{
				q.SchedulerId = "MDI scheduler";
				q.UseMicrosoftDependencyInjectionScopedJobFactory(jfo=>{
					jfo.AllowDefaultConstructor = true;
				});
			});`
and 
`services.AddQuartz(q=>{
				q.SchedulerId = "MDI scheduler";
				q.UseMicrosoftDependencyInjectionScopedJobFactory(jfo=>{
					jfo.AllowDefaultConstructor = false;
				});
			});`

**Reply by lahma** [reply]
For now (until 3.3) you also need to make sure that you have registered the job in container. This requirement will be lifted in 3.3. There's the [example ASP.NET Core project](https://github.com/quartznet/quartznet/tree/main/src/Quartz.Examples.AspNetCore) which shows a working example.

**Reply by ruza-kavanka** [reply]
Oh, now I see, thank you very much for your help Marko, I appreciate
